src/charm.py

```diff
@@ -69,29 +71,65 @@ def _on_config_changed(self, event) -> None:
         self.model.unit.status = MaintenanceStatus(msg)
         self.configure_workload_container(event)
 
+    def _get_ip_type(self, address) -> str:
+        """Validate a string type: ipv4, ipv6 or invalid as IP address."""
+        try:
+            ip = ipaddress.ip_address(address)
+
+            if isinstance(ip, ipaddress.IPv4Address):
+                return "ipv4"
+            elif isinstance(ip, ipaddress.IPv6Address):
+                return "ipv6"
+        except ValueError:
+            return
+
+    def readlines_reverse(self, qfile):
+        with qfile:
+            qfile.seek(0, os.SEEK_END)
+            position = qfile.tell()
+            line = StringIO("")
+            while position >= 0:
+                qfile.seek(position)
+                next_char = qfile.read(1)
+                if next_char == "\n":
+                    yield line.getvalue()[::-1]
+                    line = StringIO("")
+                else:
+                    line.write(next_char)
+                position -= 1
+            yield line.getvalue()[::-1]
+
     def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
-        """Handle the report-visits-by-ip action."""
-        results = self._report_visits_by_ip(event)
+        """Handle the report-visits-by-ip action.
+
+        Args:
+            event: the Juju action event fired when the action executes.
+        """
+        results = self._report_visits_by_ip()
         event.set_results({"ips": str(results)})
 
-    def _report_visits_by_ip(self, _) -> list[(int, str)]:
+    def filter_lines(self, file) -> list[str]:
```

Comment:
Maybe this would be a helpful function and we define the predicate that stops the yielding? https://docs.python.org/3/library/itertools.html#itertools.takewhile

It would also be better for this function not to take the file and instead accept the iterable with the lines of the file to reduce coupling

something like 
```Python
def stop_take(line: str) -> bool:
    line_elements = line.split()

    if len(line_elements) < 4:
        return False

    timestamp_str = line_elements[3].strip("[]")
    try:
        timestamp = datetime.strptime(timestamp_str, "%d/%b/%Y:%H:%M:%S")
    except <error if that element is not actually a timestamp>:
        return False

    return timestamp > (datetime.now() - timedelta(minutes=20))

itertools.takewhile(stop_take, reversed_lines)
```

There are still problems in the code above, e.g., is it really an exit condition if the line is empty or doesn't contain a date? Or should the line be skipped until we find a line with a date that fails the timestamp check?

Reply:
The type hint is also not quite right: https://docs.python.org/3/library/typing.html#typing.Generator